pkg/flow/reconciler/common/base.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// NewGenericServiceReconciler creates a new GenericServiceReconciler and
+// attaches a default event handler to its Service informer.
+func NewGenericServiceReconciler(ctx context.Context, gvk schema.GroupVersionKind,
+	tracker tracker.Interface,
+	adapterHandlerFn func(obj interface{}),
+) GenericServiceReconciler {
+
+	serviceinformerv1.Get(ctx).Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.FilterControllerGVK(gvk),
+		Handler:    controller.HandleAll(adapterHandlerFn),
+	})
+
+	return newGenericServiceReconciler(ctx, tracker)
+}
+
+// NewMTGenericServiceReconciler creates a new GenericServiceReconciler for a
+// multi-tenant adapter and attaches a default event handler to its Service
+// informer.
+func NewMTGenericServiceReconciler(ctx context.Context, typ kmeta.OwnerRefable,
+	tracker tracker.Interface,
+	adapterHandlerFn func(obj interface{}),
+) GenericServiceReconciler {
+
+	serviceinformerv1.Get(ctx).Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: hasAdapterLabelsForType(typ),
+		Handler:    controller.HandleAll(adapterHandlerFn),
+	})
+
+	return newGenericServiceReconciler(ctx, tracker)
+}
+
+// NewGenericRBACReconciler creates a new GenericRBACReconciler.
+func NewGenericRBACReconciler(ctx context.Context) *GenericRBACReconciler {
+	return &GenericRBACReconciler{
+		SAClient: k8sclient.Get(ctx).CoreV1().ServiceAccounts,
+		RBClient: k8sclient.Get(ctx).RbacV1().RoleBindings,
+		SALister: sainformerv1.Get(ctx).Lister().ServiceAccounts,
+		RBLister: rbinformerv1.Get(ctx).Lister().RoleBindings,
+	}
+}
+
+// newGenericServiceReconciler creates a new GenericServiceReconciler.
+func newGenericServiceReconciler(ctx context.Context,
+	tracker tracker.Interface,
+) GenericServiceReconciler {
+
+	return GenericServiceReconciler{
+		SinkResolver:          resolver.NewURIResolverFromTracker(ctx, tracker),
+		Client:                servingclient.Get(ctx).ServingV1().Services,
+		Lister:                serviceinformerv1.Get(ctx).Lister().Services,
+		GenericRBACReconciler: NewGenericRBACReconciler(ctx),
+	}
+
+}
+
+// filteredGlobalResyncFunc is a function that enqueues all objects from the
+// given informer that pass the provided filter function.
+type filteredGlobalResyncFunc func(func(interface{}) bool, cache.SharedInformer)
+
+// objectFilterFunc is a filtering function for filteredGlobalResyncFunc.
+type objectFilterFunc func(interface{}) bool
+
+// hasAdapterLabelsForType returns a function that filters based on standard
+// labels applied to all adapters of the given component type.
+func hasAdapterLabelsForType(typ kmeta.OwnerRefable) objectFilterFunc {
+	return func(obj interface{}) bool {
+		object, ok := obj.(metav1.Object)
+		if !ok {
+			return false
+		}
+
+		ls := CommonObjectLabels(typ).AsSelectorPreValidated()
+		return ls.Matches(labels.Set(object.GetLabels()))
+	}
+}
+
+// isInNamespace returns a function that filters all objects which are not in
```

Comment:
Do you mean that the comment should read "filter out" instead of "filter"?

In plain English, this is what this function is used for:
For every object found in the informer's cache, execute `isInNamespace(obj)` and keep only the ones for which the function returned true (meaning the object _is_ in the desired namespace).

I could also rephrase to
> isInNamespace returns a filter function which returns true if the object is in the given namespace.

Reply:
> Do you mean that the comment should read "filter out" instead of "filter"?

Correct. As the comment is written, I understood it to be that you were returning a function that would return object _not_ in the namespace.

Your rephrased comment reads much better.

Reply:
Done in 45306591